olc_art.hpp

```diff
@@ -84,20 +136,470 @@ class olc_db final {
+  // (multi-work) mutations to be applied to nodes.  This means that a
+  // thread can read junk in the middle of some reorganization of a node
+  // (e.g., the keys or children are being reordered to maintain an
+  // invariant for I16).  To protect against reading such junk, the thread
+  // reads the version tag before and after it accesses data in the node
+  // and restarts if the version information has changed.  The thread must
+  // not act on information that it had read until it verifies that the
+  // version tag remained unchanged across the read operation.
+  class iterator {
+    friend class olc_db;
+    template <class> friend class visitor;
+
+    // The [node_ptr] is never [nullptr] and points to the internal
+    // node or leaf for that step in the path from the root to some
+    // leaf.  For the bottom of the stack, [node_ptr] is the root.
+    // For the top of the stack, [node_ptr] is the current leaf. In
+    // the degenerate case where the tree is a single root leaf, then
+    // the stack contains just that leaf.
+    //
+    // The [key] is the [std::byte] along which the path descends from
+    // that [node_ptr].  The [key] has no meaning for a leaf.  The key
+    // byte may be used to reconstruct the full key (along with any
+    // prefix bytes in the nodes along the path).  The key byte is
+    // tracked to avoid having to search the keys of some node types
+    // (N48) when the [child_index] does not directly imply the key
+    // byte.
+    //
+    // The [child_index] is the [std::uint8_t] index position in the
+    // parent at which the [child_ptr] was found.  The [child_index]
+    // has no meaning for a leaf.  In the special case of N48, the
+    // [child_index] is the index into the [child_indexes[]].  For all
+    // other internal node types, the [child_index] is a direct index
+    // into the [children[]].  When finding the successor (or
+    // predecessor) the [child_index] needs to be interpreted
+    // according to the node type.  For N4 and N16, you just look at
+    // the next slot in the children[] to find the successor.  For
+    // N256, you look at the next non-null slot in the children[].
+    // N48 is the oddest of the node types.  For N48, you have to look
+    // at the child_indexes[], find the next mapped key value greater
+    // than the current one, and then look at its entry in the
+    // children[].
+    //
+    // Note: The read_critical_section contains the version information
+    // that must be valid to use the KB and CI data read from the NP.
+    // The CS information is cached when when those data are read from
+    // the NP along with the KB and CI values that were read.
+    using stack_entry = std::tuple
+      < detail::olc_node_ptr  // node pointer (NP)
+      , std::byte             // key byte     (KB)
+      , std::uint8_t          // child-index  (CI)
+      , version_tag           // The NP version tag invariant
```

Comment:
I don't like this pattern of unwrapping version number from a critical section and then later rehydrating (resurrecting?) new-old critical sections using this number. I believe it should be the critical section itself which should get stored here instead, with moving into stack, and moving from the stack as needed. Then `using version_tag = ...` in optimistic_lock.hpp patch drops too.

Reply:
My understanding here is incomplete ATM due to only partial review so far. Perhaps your point can be addressed by copying instead of moving RCSs then. Perhaps your current implementation is indeed the best one. We can revisit this after a full review

Reply:
I've added documentation on the version field of the stack_entry struct for olc_art.hpp.  Resolving this.